src/local-mode.ts

```diff
@@ -56,51 +99,133 @@ export class LocalModeManager {
 
 		if (!localMode) {
 			// Remove any leftovers if necessary
-			await this.removeLocalModeArtifacts();
+			await this.handleLocalModeStateChange(false);
 		}
 	}
 
-	public async removeLocalModeArtifacts(): Promise<void> {
-		try {
-			const images = await this.getLocalModeImages();
-			const containers = await this.getLocalModeContainers(images);
-
-			await Bluebird.map(containers, containerId => {
-				log.debug('Removing local mode container: ', containerId);
-				return this.docker.getContainer(containerId).remove({ force: true });
-			});
-			await Bluebird.map(images, imageId => {
-				log.debug('Removing local mode image: ', imageId);
-				return this.docker.getImage(imageId).remove({ force: true });
-			});
-
-			// Remove any local mode state added to the database
-			await this.db
-				.models('app')
-				.del()
-				.where({ source: 'local' });
-		} catch (e) {
-			log.error('There was an error clearing local mode artifacts: ', e);
-		}
+	// Query the engine to get currently running containers and installed images.
+	public async collectEngineSnapshot(): Promise<EngineSnapshotRecord> {
+		const containersPromise = this.docker
+			.listContainers()
+			.then(resp => _.map(resp, 'Id'));
+		const imagesPromise = this.docker
+			.listImages()
+			.then(resp => _.map(resp, 'Id'));
+		const volumesPromise = this.docker
+			.listVolumes()
+			.then(resp => _.map(resp.Volumes, 'Name'));
+		const networksPromise = this.docker
+			.listNetworks()
+			.then(resp => _.map(resp, 'Id'));
```

Comment:
Ah, and the only use of `then`. :)

My goal was to make sure that there is no limitation in how queries to objects are made (we don't need to make them serially), hence I moved the `await` call down to the `.all()`.

Reply:
I'd still prefer something like
```
await Bluebird.all([
  _.map(await this.docker.listContainers(), 'Id'),
...
```

but good idea on running them in parallel.

Reply:
And also unwrapping them specifically from the promise:
```typescript
const [ containers, images, volumes, networks ] = await Bluebird.all([
```

Reply:
But in construction like
```
await Bluebird.all([
  _.map(await this.docker.listContainers(), 'Id'),
  _.map(await this.docker.listImages(), 'Id'),
...
```
won't it pause until containers are collected, and only then make a request to get images?

Unwrapping is done - thanks for the hint.

Reply:
Ah yeah you're right, my bad.